lambda_src/lambda_function.py

```diff
@@ -97,15 +102,115 @@ def sync_flow(event: Any, context: Any = None) -> Dict[Text, Any]:
     LOG.debug('Destination header not found in a POST and hence using sync_flow().')
     headers = event['headers']
     req_body = loads(event['body'])
+    start_time = timer()
 
+    destination_driver = None
     batch_id = headers[BATCH_ID_HEADER]
     write_uri = headers.get('write-uri')
+
     LOG.debug(f'sync_flow() received destination: {write_uri}.')
 
     if write_uri:
         destination_driver = import_module(
             f'geff.drivers.destination_{urlparse(write_uri).scheme}'
         )
+
+    while True:
+        data = get_data_from_lock(batch_id)
+        if data is None:  # request hasn't been initialized
+            break
+        elif data != LOCKED:  # if false, the response is yet to be written
+            return data
+
+    open_lock(batch_id)  # initilaze the request
+
+    res_data = process_request(
+        req_body, headers, event, batch_id, write_uri, destination_driver
+    )
+
+    # Write data to s3 or return data synchronously
+    if write_uri:
+        response = destination_driver.finalize(  # type: ignore
+            write_uri, batch_id, res_data
+        )
+    else:
+        data_dumps = dumps({'data': res_data}, default=str)
+        response = {
+            'statusCode': 200,
+            'body': b64encode(compress(data_dumps.encode())).decode(),
+            'isBase64Encoded': True,
+            'headers': {'Content-Encoding': 'gzip'},
+        }
+        end_time = timer()
+        if response and (end_time - start_time) > 20:
+            LOG.debug('Storing the response in DynamoDB.')
+            try:
+                close_lock(batch_id, response)  # write the response
+            except ClientError as ce:
+                if ce.response['Error']['Code'] == 'ValidationException':
+                    LOG.error(ce)
+                    pass
+
+    if len(response) > 6_000_000:
+        response = response_size_error(response, req_body)
+    return response
+
+
+def response_size_error(
+    response_input: Dict[Text, Any], req_body: Dict[Text, Any]
+) -> str:
+    """
+    Creates a new response object with an error message,
+    for when the response size is likely to exceed the allowed payload size.
+
+    Args:
+        response (Dict[Text, Any]): Response object to calculate the size.
+        req_body (Any): Body of the request, obtained from the events object.
+
+    Returns:
+        Dict[Text, Any]: Represents the response with the error message.
+    """
+    response = dumps(
+        {
+            'data': [
+                [
+                    rn,
+                    {
+                        'error': (
+                            f'Response size ({len(response_input)} bytes) will likely'
+                            'exceeded maximum allowed payload size (6291556 bytes).'
+                        )
+                    },
+                ]
+                for rn, *args in req_body['data']
+            ]
+        }
+    )
+    return response
+
+
+def process_request(
```

Comment:
Functions that are called once should encapsulate meaning, so try to avoid vague functions like "process request". I'm a little confused which "request" you are referring to here... it seems misleading because of the things that might be called a "request" here, this is going to be potentially processing multiple rows per batch and each of those rows might have multiple RPC calls each of which might be represented by e.g. an HTTP request. Instead, let's call this `process_batch` and change the params a bit —

1. pull `batch_data` out of the body
2. filter out `headers` that start with `sf-custom-` and do the string replacements on them here
3. parse the `path` out of the `event`

Also, since this is only called once, you don't need to have default params that are always present.